core/src/main/java/hudson/TcpSlaveAgentListener.java

```diff
@@ -384,6 +429,65 @@ public boolean connect(Socket socket) throws IOException {
     }
 
     /**
+     * Reschedules the <code>TcpSlaveAgentListener</code> on demand.  Disables itself after running.
+     */
+    @Extension
+    public static class TcpSlaveAgentListenerRescheduler extends AperiodicWork {
+        private Thread originThread;
+        private int port;
+        private Throwable cause;
+        private long recurrencePeriod = 5000;
+        private boolean isActive;
+
+        public TcpSlaveAgentListenerRescheduler(Thread originThread, int port, Throwable cause) {
+            this.originThread = originThread;
+            this.port = port;
+            this.cause = cause;
+            this.isActive = false;
+        }
+
+        @Override
+        public long getRecurrencePeriod() {
+            return recurrencePeriod;
+        }
+
+        @Override
+        public AperiodicWork getNewInstance() {
+            return new TcpSlaveAgentListenerRescheduler(originThread, port, cause);
+        }
+
+        @Override
+        protected void doAperiodicRun() {
+            if (isActive) {
+                try {
+                    if (originThread.isAlive()) {
+                        originThread.interrupt();
+                    }
+                    new TcpSlaveAgentListener(port).start();
+                    LOGGER.log(Level.INFO, "Restarted TcpSlaveAgentListener");
+                } catch (IOException e1) {
+                    LOGGER.log(Level.SEVERE, "Could not reschedule TcpSlaveAgentListener", cause);
```

Comment:
Reschedule another attempt? If the Agent port is enabled, there is no valid use-case for the dead thread. It would also indicate that there is a pending problem if the thread is alive and rejects to stop

Reply:
I moved `isActive = false` to only happen if the thread is successfully restarted.  I think that means that it will automatically retry?